apollo-router/src/spec/field_type.rs

```diff
@@ -31,14 +31,18 @@ impl Serialize for FieldType {
             {
                 #[derive(Serialize)]
                 enum NestedBorrowed<'a> {
-                    NonNull(BorrowedFieldType<'a>),
-                    List(BorrowedFieldType<'a>),
                     Named(&'a str),
+                    NonNullNamed(&'a str),
+                    List(BorrowedFieldType<'a>),
+                    NonNullList(BorrowedFieldType<'a>),
```

Comment:
Hmm these types aren't quite inline with the spec. Is it just to make it easier with the borrows?

Reply:
Having an ad-hoc rust enum here at all is to deal with borrows. The change in variants is for a corresponding one [in the compiler 1.0 AST](https://docs.rs/apollo-compiler/1.0.0-beta.4/apollo_compiler/ast/enum.Type.html):

```rust
pub enum Type {
    Named(NamedType),
    NonNullNamed(NamedType),
    List(Box<Type>),
    NonNullList(Box<Type>),
}
```

This representation is more in line with the spec than compiler 0.11 which allowed non-null of non-null. The spec excludes non-null of non-null [in its grammar](https://spec.graphql.org/October2021/#Type):

* Type:
  -  NamedType
  - ListType
  - NonNullType
* NamedType:
  - Name
* ListType:
  - `[` Type `]`
* NonNullType
  - NamedType `!`
  - ListType `!`

Which simplifies to:

* Type:
  - Name
  - `[` Type `]`
  - Name `!`
  - `[` Type `]` `!`